### PR TITLE
Fix rust layer call to `rustup`

### PR DESCRIPTION
**Pull request by Dietr1ch**
Bash compatibility was assumed, which made the shell command to fail on
fish and probably other shells too.

### CODE REVIEW

**Comment by Dietr1ch** [comment]
I don't know, but I'd bet for it.
I think there's no single commandline that runs " A && B" on arbitrary shells and systems.
Do you know how to do that `and` on elisp rather than relying on the shell command?

**Comment by joehillen** [comment]
I think what you'd want is to check the result of:

```elisp
(compile
  (format "rustc -o %s %s"
          (shell-quote-argument output-file-name)
          (shell-quote-argument input-file-name)))
```

and run `(compile (shell-quote-argument output-file-name))` if it passes.

**Comment by syl20bnr** [comment]
@Dietr1ch what about @joehillen comment ? Do you plan to integrate it ?

**Comment by Dietr1ch** [comment]
I can, but I'm not sure how to test the output of that `compile`

I thought that it would be simply,
```
(and
 (compile
  (format "rustc -o %s %s"
          (shell-quote-argument output-file-name)
          (shell-quote-argument input-file-name)))
 (compile (shell-quote-argument output-file-name))
 )
```

I checked `compile` docs with `SPC h d f compile`, but they don't mention how to get the status code.

**Comment by Dietr1ch** [comment]
It still needs some work, it's not catching compile errors :/

**Comment by Dietr1ch** [comment]
There was a silly "\n" at the end of the status.., now it works

**Comment by Miciah** [comment]
Just noticed this PR.  https://github.com/syl20bnr/spacemacs/pull/10610 is a different solution to the same problem (namely, the other PR runs the command using `sh -c`; simpler, not sure about portability of using `sh -c` though).  Which solution is preferred?

File: layers/+lang/rust/funcs.el [hunk 1]
```diff
@@ -24,10 +24,28 @@ open a junk Rust file, type in some code and quickly run it.
...
+                    (md5 (buffer-file-name))))
+                )
+    (add-to-list 'compilation-finish-functions
+                 (lambda (buffer status)
+                   (if (string-match "finished" status)
+                     (shell-command
+                       (shell-quote-argument output-file-name)
+                       buffer
+                       )
+                     (message "Complation failed with: %s" status)
```
Comment (Miciah) [inline layers/+lang/rust/funcs.el]: Typo: s/Complation/Compilation/

**Comment by Dietr1ch** [comment]
@Miciah, this started as the same `sh -c` hack, but windows compatibility was uncertain.
I consider moving to lisp is better because you can actually distinguish the build from the run.

I'll fix the hanging parentheses in the hopes that I get used to them at some point (maybe I just
need to learn how to really edit S-expressions on emacs rather than keeping my vi habits)

**Comment by Dietr1ch** [comment]
The failure at `ci/circleci: core` is an infra failure,

> Starting container jare/spacemacs-circleci:emacs24
>   image cache not found on this host, downloading jare/spacemacs-circleci:emacs24
> Error response from daemon: manifest for jare/spacemacs-circleci:emacs24 not found

**Comment by syl20bnr** [comment]
See comment: https://github.com/syl20bnr/spacemacs/pull/10610#issuecomment-397785479

**Comment by Dietr1ch** [comment]
IIRC this issue was becaue using `&&` to chain commands doesn't work on fish (I think there's a recent hack/change to allow it).
Grabbing the right environment is nice as it ensures consistently with your shell, but it won't make the `&&` work.

The issue with using`sh -c "a && b"` was suspected incompatibility with windows, if it actually works it may be an ok solution whether we care or not about mixing the outputs or not.

**Comment by syl20bnr** [comment]
What about using `shell-command-to-string` which should use the shell
pointed by `shell-file-name` ?

**Comment by smile13241324** [comment]
Thank you for contributing to spacemacs 💜, I have cherry picked your commit into develop.

**Comment by duianto** [comment]
I haven't tested this updated `rust` function, but I noticed that it added a: `lexical-let`

The `cl` package was recently deprecated in Spacemacs, in favor of: `cl-lib`

It caused an issue with the `ivy` layer that used `lexical-let`.
It was fixed by this commit:
Remove uses of lexical-let, part of the deprecated cl package https://github.com/syl20bnr/spacemacs/commit/05fb56bfbfd43a38d0b4044796d2cff4cccac971

I'm just wondering if a similar issue might occur in the `rust` layer or at least with the usage of this function?

**Pull request closed without merging**